src/ir/datalog_print_context.h

```diff
@@ -26,7 +28,7 @@ namespace raksha::ir {
 // unique labels.
 class DatalogPrintContext {
  public:
-  DatalogPrintContext() : check_counter_(0) {}
+  DatalogPrintContext() : check_counter_(0), instantiation_map_(nullptr) {}
```

Comment:
Rather than setting the instantiation map to nullptr initially then resetting it for each particle instance, it might be nicer to wrap the `check_counter_` in its own simple `CheckCounter` class. A pointer to `CheckCounter` can then be passed in in the ctor alongside the instantiation map. Then, a new `DatalogPrintContext` can be constructed for each new `Particle`, preventing `instantiation_map_` from ever being `nullptr` and removing one piece of mutable state.

That said, I hear your point about this being a temporary fix until we have a full-fledged visitor, so I'm not strongly opposed to the pattern you have here right now.

Reply:
If you are worried about `instantiation_map_` from being null, we could fix it by making `instantiation_map_` mandatory and the `check_counter` an optional argument. We can reinitialize the counter between particles. We will have to do something like that with `CheckCounter` as well.

In any case, let us leave this as is for now. When we have a visitor and the core data structures, this will become cleaner.